Add bulk publish documentation (#3005)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/pubsub/pubsub-bulk.md

---
type: docs
title: "Send and receive messages in bulk"
linkTitle: "Send and receive messages in bulk"
weight: 7100
description: "Learn how to use the bulk publish and subscribe APIs in Dapr."
---

{{% alert title="alpha" color="warning" %}}
The bulk publish and subscribe APIs are in **alpha** stage.
{{% /alert %}}

With the bulk publish and subscribe APIs, you can send and receive multiple messages in a single request.

## Native bulk publish and subscribe support

When a pub/sub component supports the bulk publish API natively, Dapr also publishes messages to the underlying pub/sub component in bulk.

Otherwise, Dapr falls back to sending messages one by one to the underlying pub/sub component. This is still more efficient than using the regular publish API, because applications can still send multiple messages in a single request to Dapr.

## Supported components

> Refer [component reference]({{< ref supported-pubsub >}}) to see which components support the bulk publish API natively.

Reviewer: ```suggestion
Refer to the [component reference]({{< ref supported-pubsub >}}) to see which components support the bulk publish API natively.
```

---
type: docs
title: "Send and receive messages in bulk"

> linkTitle: "Send and receive messages in bulk"

Reviewer 2: Sometime you have "send and receive" and other times you have "publish and subscribe" in this article. I would pick "publish and subscribe" for consistency.

weight: 7100
description: "Learn how to use the bulk publish and subscribe APIs in Dapr."
---

{{% alert title="alpha" color="warning" %}}
The bulk publish and subscribe APIs are in **alpha** stage.
{{% /alert %}}

> With the bulk publish and subscribe APIs, you can send and receive multiple messages in a single request.

Reviewer 2: It is always good to cover in the intro, the "why" you would use this feature. This topic dives straight into describing bulk pub/sub and giving a list of common scenario would make the reader understand if they need to continue, or this was not needed.

Author: I have added a one-liner here, the gist to the get high performance if the number of messages/events is too high, do you think we should add something else?

> ## Native bulk publish and subscribe support

Reviewer 2: This section should be moved to later in the document. It is strange to encounter this as the first section to read about.
Also "native" here is not explained. Instead I would open with "Some pub/sub brokers have support for publishing and receiving multiple "bulk" messages, which can be referred to as native support. When a component supports this, then the Dapr runtime uses the component native capability . For components that do not have this bulk message support, Dapr runtime sends messages one by one to the pub/sub component. This is more efficient than allowing an  application to send individual message as the application sends multiple messages in a single request to Dapr.

Reviewer 2: Also I would change the title to 
```suggestion
## How components handle sending bulk publish and subscribe messages
```

Author: > This section should be moved to later in the document. It is strange to encounter this as the first section to read about.

Moved it after the bulk publish example. Removed the "native" terminology.

When a pub/sub component supports the bulk publish API natively, Dapr also publishes messages to the underlying pub/sub component in bulk.

Otherwise, Dapr falls back to sending messages one by one to the underlying pub/sub component. This is still more efficient than using the regular publish API, because applications can still send multiple messages in a single request to Dapr.

## Supported components

> Refer to the [component reference]({{< ref supported-pubsub >}}) to see which components support the bulk publish API natively.

Reviewer 2: Little confused here.  Given the above, why do I care if the component supports this "natively" if I can do this with any pub/sub component, correct? Or can the pub/sub API **only** be used with these components. If this is only for the ones supporting it, then I would not use the term "native" as this is unnecessary.  Can you clarify which pub/sub components can use bulk (all of them or just the ones with this "natively") and if one some of them, then I would remove the term "native" altogether and simply say they support bulk send/receive or not.

Author: Right, I am re-wording this to remove "native" and just say whether they support bulk or not. So as of now, we are telling users to use the bulk API to optimize app<==>sidecar communication. Further, if a component supports bulk ops, we can also optimize sidecar<==>broker communication. There is no need for users to know the internal plumbing of Dapr and learn the "native"/"default" terminologies.

> ## Publishing messages in bulk

Reviewer 2: ```suggestion
## Restrictions when publishing messages in bulk
```

> The bulk publish API allows you to publish multiple messages to a topic in a single request. If any of the messages fail to publish, the bulk publish operation returns a list of failed messages. Note, the bulk publish operation does not guarantee the order of messages.

Reviewer 2: ```suggestion
The bulk publish API allows you to publish multiple messages to a topic in a single request. Some of these message succeed and some of them can fail. If any of the messages fail to publish, the bulk publish operation returns a list of failed messages. 

The bulk publish operation also does not guarantee any ordering of messages.
```

Reviewer 2: To be clear on the publish, this is not transactional, in that they all succeed or or fail. If that is true, then I would use the term "non-transactional" to make this clear

Author: Right, added the term.

### Example

> {{< tabs Java Javascript "HTTP API (Bash)" "HTTP API (PowerShell)" >}}

Reviewer 2: You need to add all supported language
.NET, Go, Python. Suggest asking for help if needed

Author: There are only three SDKs with this support (added dotnet also now). Followed our existing docs where we usually give a plain HTTP example + SDK examples https://docs.dapr.io/developing-applications/building-blocks/distributed-lock/howto-use-distributed-lock/#acquire-lock

Adding Python/Go with simple HTTP clients.

    // Publish multiple messages to a topic.
    await client.pubsub.publishBulk(pubSubName, topic, ["message 1", "message 2", "message 3"]);

    // Publish multiple messages to a topic with explicit bulk publish messages.
    const bulkPublishMessages = [
    {
      entryID: "entry-1",
      contentType: "application/json",
      event: { hello: "foo message 1" },
    },
    {
      entryID: "entry-2",
      contentType: "application/cloudevents+json",
      event: { 
        specversion: "1.0",
        source: "/some/source",
        type: "example",
        id: "1234", 
        data: "foo message 2", 
        datacontenttype: "text/plain" 
      },
    },
    {
      entryID: "entry-3",
      contentType: "text/plain",
      event: "foo message 3",
    },
  ];
  await client.pubsub.publishBulk(pubSubName, topic, bulkPublishMessages);
}

start().catch((e) => {
  console.error(e);
  process.exit(1);
});
```

{{% /codetab %}}

{{% codetab %}}

```bash
curl -X POST http://localhost:3500/v1.0-alpha1/publish/bulk/my-pubsub-name/topic-a \
  -H 'Content-Type: application/json' \
  -d '[
        {
            "entryId": "ae6bf7c6-4af2-11ed-b878-0242ac120002",

>             "event":  "first",

Reviewer 2: ```suggestion
            "event":  "first text message",
```

            "contentType": "text/plain"
        },
        {
            "entryId": "b1f40bd6-4af2-11ed-b878-0242ac120002",
            "event":  {

>                 "message": "second"   

Reviewer 2: ```suggestion
                "message": "second JSON message"   
```

            },
            "contentType": "application/json"
        },
      ]'
```

{{% /codetab %}}

{{% codetab %}}

```powershell
Invoke-RestMethod -Method Post -ContentType 'application/json' -Uri 'http://localhost:3500/v1.0-alpha1/publish/bulk/my-pubsub-name/topic-a' `
-Body '[
        {
            "entryId": "ae6bf7c6-4af2-11ed-b878-0242ac120002",

>             "event":  "first",

Reviewer 2: ```suggestion
            "event":  "first text message",
```

            "contentType": "text/plain"
        },
        {
            "entryId": "b1f40bd6-4af2-11ed-b878-0242ac120002",
            "event":  {

>                 "message": "second"   

Reviewer 2: ```suggestion
                "message": "second JSON message"   
```
